### PR TITLE
WIP: Bidirectional dinic

### PR DESCRIPTION
**Pull request by chistopher**
This pull request is to collect feedback to these changes and have a place to write down TODOs and further ideas.

TODO:
- [ ] benchmark flow algo on larger instances
- [ ] benchmark usage in cutter framework
- [ ] plot search space w/ and w/o bidirectional search (possibly divided in `buildLayeredNetwork` and `augmentFlowInLayeredNetwork`

### CODE REVIEW

File: algorithm/dinic_bidirectional.h [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+										scan_for_backward[e] = STATE::FLOW_SENDING;
+									}
+									
+									for (const Pin& pv : hg.pinsNotReceivingFlowFrom(e)) {
+										backward_visit(pv.pin);
+									}
+									
+								}
+								if (!are_flow_receiving_pins_target_reachable(e)) {
+									// u <- in-node(e) <- all pins sending flow into e
```
Comment (chistopher) [inline algorithm/dinic_bidirectional.h]: this comment should be `u <- out-node(e) <- all pins receiving flow from e`

File: algorithm/dinic_bidirectional.h [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+		
+		bool buildLayeredNetwork(CutterState<Type>& cs, const bool augment_flow, const bool only_forward) {
+			unused(augment_flow);	//for debug builds only
+			//cs.clearForSearch();
+			assert(cs.currentViewDirection() == 0);
+			
+			if (scan_for_backward.size() != hg.numHyperedges()) {
+				scan_for_backward.resize(hg.numHyperedges(), STATE::NONE);
+			}
+			std::fill(scan_for_backward.begin(), scan_for_backward.end(), STATE::NONE);
```
Comment (chistopher) [inline algorithm/dinic_bidirectional.h]: this should be lazy

**Comment by chistopher** [comment]
an idea that could help to simplify the dfs and other operations on the layered network: 
After both searches met, one can continue only the forward search, but restrict it to the search space of the backward search.
This allows to build all distance-labels and data structures as for the uni-directional variant and still have the benefits of bidirectional search.

File: algorithm/dinic_bidirectional.h (hunk 2 above)
Comment (larsgottesbueren) [reply algorithm/dinic_bidirectional.h]: Yup. I'm still looking for a way to get rid of this array altogether

**Pull request closed without merging**